### ISSUE TITLE
Determine the behavior when calling unobserve/disconnect method before the promise returned by observe method is settled

### ISSUE THREAD

**Issue by wangw-1991** [WICG/compute-pressure#164]
The following codes will cause the promise never resolve or reject in current implementation. It is more reasonable to reject these promises instead and the spec should give more clarity about this.

let promise = observer.observe('cpu');
observer.disconnect();

**Comment by rakuco** [comment]
More generally speaking, `PressureObserver.observe()` returns a new Promise every time it is invoked, and `PressureObserver.{disconnect,unobserve}()` may be called while `observe()` is in the "in parallel" steps. It is not clear what should be done in this case as they should probably influence what happens to the unfulfilled promises.

**Comment by kenchris** [comment]
Maybe this would work:

- store pending observe promises
- disconnect: reject all pending observe promises, but make sure that everything has been removed from "active observer list" - that should be fine with current spec text
- unobserve: iterate over the pending promises and reject the ones matching the source. Remove from "active observer list" - might have been added, might not

Biggest issue I see if the "global task" as that will run when after JS has completed, including the sync disconnect/unobserve, so it will readd to the active observer list and activate the observation, so that code would probably have to check whether it's promise has been removed from the pending observer promises list and if so, don't do any work

**Comment by kenchris** [comment]
Given it looks like we will remove the async permission stuff, this becomes a non-issue. But we should keep using promises in observe() so that we can add it back.

**Comment by rakuco** [comment]
Given the current Blink implementation, I think this will still be an issue even if the permission bits are removed, as there are async steps involved in checking if `source` is a supported source type and when activating data delivery.

It could be possible to stop distinguishing between a successful async connection to the platform side (and maybe even drop the NotSupportedError exception related to checking for supported source types), but one would still have a window where the async connection is being established internally but disconnect/unobserve can still be called.

**Comment by rakuco** [comment]
For the record, at least Web Bluetooth and WebHID also need to deal with this:
* https://webbluetoothcg.github.io/web-bluetooth/#dom-bluetooth-attributeinstancemap-slot
* https://wicg.github.io/webhid/#dfn-pendingsendreportpromises
* https://wicg.github.io/webhid/#dfn-pendingsendfeaturereportpromises
* https://wicg.github.io/webhid/#dfn-pendingreceivefeaturereportpromises

**Comment by kenchris** [comment]
Given that we use an attribute for supported source types, we have not made that an async querying function, so we could do this check sync and it should not be an issue.

If we want to make that async in the future, it does become a problem yes.

**Comment by kenchris** [comment]
If that is indeed async, how is the attribute implemented today then? Also the current activation was happening on the main tread (in the global task) so the promise resolve wasn't waiting on that, you just get the events when they are delivered

**Comment by kenchris** [comment]
> one would still have a window where the async connection is being established internally but disconnect/unobserve can still be called.

We haven't really spec'ed how that happens, and in the spec that also happens in the global task (main thread) so if we want to do that, we need to find out how to make these changes

**Comment by kenchris** [comment]
@rakuco I did a PR to remove the permissions integration, but I can change it to keep the global task and instead do the activation in parallel. Then afterwards we could do the pendingPromises. If you think that is the way to go, tell me and I will prepare the PRs

**Comment by kenchris** [comment]
@arskama mentioned that the attribute getter returns only values supported by the spec and not necessarily the underlying hardware - I believe that is how it worked in some other specs, we can consider whether to change that or not.

**Comment by rakuco** [comment]
> @arskama mentioned that the attribute getter returns only values supported by the spec and not necessarily the underlying hardware

Values supported by the spec or by the implementation? If it's the former, then `PressureSource` suffices; if it's the latter, then it's either what the platform supports (which you only know asynchronously) or it's a third state that means "supported by the implementation but not necessarily by the platform" (in which case `supportedSources` does not necessarily return the same values as those used in the observe() check).

**Comment by kenchris** [comment]
Basically what the browser supports on at least one of its platforms

**Comment by kenchris** [comment]
Btw looking at those WebHID pending promises, I don't ever see them remove promises as they resolve. Looks like a mistake, or am I missing something?

**Comment by arskama** [comment]
https://wicg.github.io/compute-pressure/#platform-primitives
As written in the section mentioned above. The platform collector is the SW proxy to access the HW,which means that it can be all synchronously requested.

The question remaining is: How do we inform the user on the case where the HW doesn't support the platform collector requests? If observe is synchronous, there is no way to do it.

**Comment by kenchris** [comment]
@Elchi3 I was wondering if you have any input on what supportedSources should expose. Right now it exposes the sources supported by the user agent (browser) but not necessarily supported on a platform (say Android) or by the hardware.

Should we change that and make it an async getter, or add a similar method with that behavior in addition to the attribute?

**Comment by kenchris** [comment]
Yeah so we could add a ```Promise<FrozenArray<string>> getSupportedSources()``` in place of the existing method

I think that would be more useful. @rakuco @arskama what do you think? How hard would this be @arskama ?

**Comment by kenchris** [comment]
If we decide to do this, we can remove the existing API for now and if we don't make it, we can add this new method after shipping

@tomayac you feedback is also welcome

**Comment by arskama** [comment]
Before M125 release, I don't see it happening.

So if we want to provide the platform collector source list, then it s maybe better to remove it now and add it later when we have an implementation.
